### DISCUSSION TITLE
Logging request and response body with RequestLogFactory

### DISCUSSION THREAD

**Discussion by Kaban77** [dropwizard/dropwizard#8227] [Q&A]
I'd like to log request and response body. 
I tried to implement interface _io.dropwizard.request.logging.RequestLogFactory_ and extend _ch.qos.logback.access.jetty.RequestLogImpl_ (method _log(Request jettyRequest, Response jettyResponse)_) but input steam and output stream were already read.
How can I log request and response body?

**Comment by mattnelson** [comment]
I haven't tried, so can't vouch, but check these logback docs on it.

https://logback.qos.ch/recipes/captureHttp.html

**Comment by zUniQueX** [comment]
For completeness sake:

The documentation kindly provided by @mattnelson used configurations with `web.xml` and `logback-access.xml`. In Dropwizard we don't use the `web.xml` or the `logback-access.xml`. To use the `TeeFilter` you would do something like the following:
 1. Register the `TeeFilter`: To register the `TeeFilter` you'd add the following lines to your application's run method:
    ```java
    environment.servlets().addFilter("TeeFilter", ch.qos.logback.access.servlet.TeeFilter.class)
               .addMappingForUrlPatterns(EnumSet.allOf(DispatcherType.class), false, "/*");
    ```
 2. Write a custom logging layout: Dropwizard only has a predefined layout for the 'normal' use case. To use the conversion words `%fullRequest` and `%fullResponse` explained in the docs, you need to provide a layout using those conversion words. The layout could look like this:
    ```java
    @JsonTypeName("full-logback-access")
    public class FullAccessLogLayoutFactory implements DiscoverableLayoutFactory<IAccessEvent> {

        @Override
        public LayoutBase<IAccessEvent> build(LoggerContext context, TimeZone timeZone) {
            PatternLayout patternLayout = new PatternLayout();
            patternLayout.setPattern("%fullRequest%n%n%fullResponse");
            patternLayout.setContext(context);
            return patternLayout;
        }
    }
    ```
 3. Make the layout discoverable: To use the layout in your application's config file, the layout must be discoverable by Jackson. To do this you'll create a file `src/main/resources/META-INF/services/io.dropwizard.logging.common.layout.DiscoverableLayoutFactory` with the content:
    ```
    your.package.to.FullAccessLogLayoutFactory
    ```
 4. Use the layout: To use the layout, you'll then override the default request logging configuration:
    ```yaml
    server:
      requestLog:
        appenders:
          - type: console
            layout:
              type: full-logback-access
    ```

Please note: this should be done only for testing purposes. The `TeeFilter` wraps the streams in a `ByteArrayOutputStream` and `ByteArrayInputStream`. This holds the entire request/response in memory, which is contrary to streaming input and output e.g. with chunked transfer enconding.
